### PR TITLE
Implements quantity system

### PR DESCRIPTION
**Pull request by empireu**
Generalizes physical quantities by introducing units and unit scales. Also replaces `Temperature`

### CODE REVIEW

File: src/main/kotlin/org/ageseries/libage/data/Quantity.kt [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+    /**
+     * Amplifies this scale 1000 times.
+     * */
+    operator fun unaryPlus() = this * 1000.0
+
+    /**
+     * Reduces this scale 1000 times.
+     * */
+    operator fun unaryMinus() = this / 1000.0
```
Comment (Grissess) [inline src/main/kotlin/org/ageseries/libage/data/Quantity.kt:33]: While I get it for SI purposes, these implementations _as an operator_ are a little dubious; it won't stop me from merging this beta software, but I can't see this being useful as opposed to, say, a dedicated method

File: src/main/kotlin/org/ageseries/libage/data/Quantity.kt [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    operator fun unaryMinus() = Quantity<Unit>(-value)
+    operator fun unaryPlus() = Quantity<Unit>(+value)
+    operator fun plus(b: Quantity<Unit>) = Quantity<Unit>(this.value + b.value)
+    operator fun minus(b: Quantity<Unit>) = Quantity<Unit>(this.value - b.value)
+    operator fun times(scalar: Double) = Quantity<Unit>(this.value * scalar)
+    operator fun div(scalar: Double) = Quantity<Unit>(this.value / scalar)
+    /**
+     * Divides the quantity by another quantity of the same unit. This, in turn, cancels out the quantity, returning the resulting number.
+     * */
+    operator fun div(b: Quantity<Unit>) = this.value / b.value
```
Comment (Grissess) [inline src/main/kotlin/org/ageseries/libage/data/Quantity.kt:83]: Neat! I'm tempted to do a little more dimensional analysis, too (have Unit itself be a bag of basic units with exponents), but that's probably out of scope.

File: src/main/kotlin/org/ageseries/libage/data/Quantity.kt [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+fun <U> max(a: Quantity<U>, b: Quantity<U>) = Quantity<U>(kotlin.math.max(!a, !b))
+fun <U> abs(q: Quantity<U>) = Quantity<U>(kotlin.math.abs(!q))
+
+/**
+ * Defines the standard scale of the [Unit] (a scale with factor 1).
+ * */
+fun <Unit> standardScale(factor: Double = 1.0, base: Double = 0.0) = QuantityScale<Unit>(factor, base)
+
+val KILOGRAMS = standardScale<Mass>()
+val GRAMS = -KILOGRAMS
```
Comment (Grissess) [inline src/main/kotlin/org/ageseries/libage/data/Quantity.kt:109]: I'm picking up what you're putting down with these, but I still think something like `.milli` and `.kilo` would still read better :P

File: src/main/kotlin/org/ageseries/libage/data/Quantity.kt [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+val JOULE = standardScale<Energy>()
+val KILOJOULES = +JOULE
+val MEGAJOULES = +KILOJOULES
+val GIGAJOULES = +MEGAJOULES
+val WATT_SECONDS = QuantityScale<Energy>(JOULE.factor, 0.0)
+val WATT_MINUTES = WATT_SECONDS * 60.0
+val WATT_HOURS = WATT_MINUTES * 60.0
+val KW_HOURS = WATT_HOURS * 1000.0
+
+// Serious precision issues? Hope not! :Fish_Smug:
```
Comment (Grissess) [inline src/main/kotlin/org/ageseries/libage/data/Quantity.kt:132]: You're so lucky GitHub doesn't have my numerous dragon emoji _yet_

fisheyes

File: src/main/kotlin/org/ageseries/libage/data/Quantity.kt [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+val ELECTRON_VOLT = JOULE * 1.602176634e-19
+val KILO_ELECTRON_VOLT = JOULE * 1.602176634e-16
+val MEGA_ELECTRON_VOLT = JOULE * 1.602176634e-13
+val GIGA_ELECTRON_VOLT = JOULE * 1.602176634e-10
+val TERA_ELECTRON_VOLT = JOULE * 1.602176634e-7
```
Comment (Grissess) [inline src/main/kotlin/org/ageseries/libage/data/Quantity.kt:137]: `.kilo`?

File: src/main/kotlin/org/ageseries/libage/data/Quantity.kt [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+val MEGABECQUERELS = +KILOBECQUERELS
+val GIGABECQUERELS = +MEGABECQUERELS
+val TERABECQUERELS = +GIGABECQUERELS
+val CURIE = GIGABECQUERELS * 37.0
+val MILLICURIES = MEGABECQUERELS * 37.0
+val MICROCURIES = KILOBECQUERELS * 37.0
+val NANOCURIES = BECQUEREL * 37.0
+val KILOCURIES = +CURIE
+val MEGACURIES = +KILOCURIES
+val GIGACURIES = +MEGACURIES // Average conversation with Grissess (every disintegration is a cute dragon image)
```
Comment (Grissess) [inline src/main/kotlin/org/ageseries/libage/data/Quantity.kt:150]: dragontired

**Review by Grissess — APPROVED** [review]
Some changes are a good idea, but I'm not gonna _require_ their address :)

**Comment by Grissess** [comment]
begone codacy

File: src/main/kotlin/org/ageseries/libage/data/Quantity.kt (hunk 4 above)
Comment (jrddunbr) [reply src/main/kotlin/org/ageseries/libage/data/Quantity.kt:132]: Yet.....

**Comment by jrddunbr** [comment]
my two cents on this PR despite being merged - the number of constants is dizzying and really should focus on building that into the actual code structure instead of building a massive array of hand-managed constants.

I like the thought of having .kilo and .mega and such on appropriate types.

**Pull request merged**